src/python/pants/engine/target.py

```diff
@@ -0,0 +1,193 @@
+
+        The value will possibly be first hydrated and/or validated, such as using a default value
+        if the field was not defined or ensuring that an int value is positive.
+
+        This property is memoized because hydration and validation can often be costly. This
+        hydration is lazy, i.e. it will only happen when a downstream rule explicitly requests this
+        field.
+        """
+
+
+class AsyncField(Field, metaclass=ABCMeta):
+    """A field that needs the engine in order to be hydrated.
+
+    You should create a corresponding Result class and define a rule to go from this AsyncField to
+    the Result. For example:
+
+        class Sources(AsyncField):
+            alias: ClassVar = "sources"
+            raw_value: Optional[List[str]]
+
+        @dataclass(frozen=True)
+        class SourcesResult:
+            snapshot: Snapshot
+
+
+        @rule
+        def hydrate_sources(sources: Sources) -> SourcesResult:
+            sources.validate_pre_hydration()
+            result = await Get[Snapshot](PathGlobs(sources.raw_value))
+            sources.validate_post_hydration()
+            return SourcesResult(result)
+
+
+        def rules():
+            return [hydrate_sources]
+
+    Then, call sites can `await Get` if they need to hydrate the field:
+
+        sources = await Get[SourcesResult](Sources, my_tgt.get(Sources))
+    """
+
+    def validate_pre_hydration(self) -> None:
+        """Any validation that can be done on the original `raw_value`.
+
+        It is cheaper to do any possible validation here, rather than in `validate_post_hydration`,
+        because we can short-circuit if an invariant is violated before incurring the expense of
+        hydration.
+        """
+
+    def validate_post_hydration(self, result: Any) -> None:
+        """Any validation that must be done after hydration by inspecting the result of that
+        hydration.
+
+        For example, a `PythonSources` field may validate that all hydrated source files end in
+        `.py`.
+        """
+
+
+class PluginField:
+    """Allows plugin authors to add new fields to pre-existing target types via UnionRules.
+
+    When defining a Target, authors should create a corresponding PluginField class marked with
+    `@union`. Then, plugin authors simply need to create whatever new `Field` they want and in a
+    `register.py`'s `rules()` function, call `UnionRule`. For example, to add a
+    `TypeChecked` field to `python_library`, register `UnionRule(PythonLibraryField, TypeChecked)`.
+
+        @union
+        class PythonLibraryField(PluginField):
+            pass
+
+
+        class PythonLibrary(Target):
+            core_fields = (Compatibility, PythonSources, ...)
+            plugin_field_type = PythonLibraryField
+
+
+        class TypeChecked(PrimitiveField):
+            ...
+
+
+        def rules():
+            return [UnionRule(PythonLibraryField, TypeChecked)]
+    """
+
+
+_F = TypeVar("_F", bound=Field)
```

Comment:
This line is important. It's what gets MyPy to understand everything, like that `my_tgt.get(Compatibility).value` would be `List[str]` but `my_tgt.get(ZipSafe).value` would be `bool`.

Reply:
That is a magical thing, but also this line is totally inscrutable. Maybe add this comment to the code?

Reply:
Eh, I think probably not to adding a comment. We use TypeVars pretty liberally in Pants and MyPy has good documentation on it: https://mypy.readthedocs.io/en/stable/generics.html

Our usage here is very idiomatic for how you do generics in Python.

I only wanted to call attention because this one line of code has much more power than you might think at first glance.